src/test/transaction_tests.cpp

```diff
@@ -94,6 +100,65 @@ std::string FormatScriptFlags(unsigned int flags)
     return ret.substr(0, ret.size() - 1);
 }
 
+bool CheckTxScripts(const CTransaction& tx, const std::map<COutPoint, CScript>& mapprevOutScriptPubKeys, const std::map<COutPoint, int64_t>& mapprevOutValues, const unsigned int& flags, const PrecomputedTransactionData& txdata, const std::string& strTest, const bool expect_valid)
+{
+    bool tx_valid = true;
+    ScriptError err = expect_valid ? SCRIPT_ERR_UNKNOWN_ERROR : SCRIPT_ERR_OK;
+    for (unsigned int i = 0; i < tx.vin.size() && tx_valid; i++)
+    {
+        const CAmount amount = mapprevOutValues.count(tx.vin[i].prevout) ? mapprevOutValues.at(tx.vin[i].prevout) : 0;
+        try
+        {
+            tx_valid = VerifyScript(tx.vin[i].scriptSig, mapprevOutScriptPubKeys.at(tx.vin[i].prevout), &tx.vin[i].scriptWitness, flags, TransactionSignatureChecker(&tx, i, amount, txdata), &err);
+        }
+        catch (...)
+        {
+            BOOST_ERROR("Bad test: " << strTest);
+            return true; // The test format is bad and an error is thrown. Return true to silence further error.
+        }
+        if (expect_valid) {
+            BOOST_CHECK_MESSAGE(tx_valid, strTest);
+            BOOST_CHECK_MESSAGE((err == SCRIPT_ERR_OK), ScriptErrorString(err));
+            err = SCRIPT_ERR_UNKNOWN_ERROR;
+        }
+    }
+    if (!expect_valid) {
+        BOOST_CHECK_MESSAGE(!tx_valid, strTest);
+        BOOST_CHECK_MESSAGE((err != SCRIPT_ERR_OK), ScriptErrorString(err));
+    }
+    return (tx_valid == expect_valid);
+}
+
+/*
+ * Trim or fill flags to make the combination valid:
+ * ** WITNESS must be used with P2SH
+ * ** CLEANSTACK must be used WITNESS and P2SH
+ * */
+
+unsigned int TrimFlags(unsigned int flags)
+{
+    if (!(flags & SCRIPT_VERIFY_WITNESS)) {
+        flags &= ~(unsigned int)SCRIPT_VERIFY_CLEANSTACK;
+    }
+    else if (!(flags & SCRIPT_VERIFY_P2SH)) {
+        flags &= ~(unsigned int)SCRIPT_VERIFY_WITNESS;
+        flags &= ~(unsigned int)SCRIPT_VERIFY_CLEANSTACK;
+    }
+    return flags;
+}
+
+unsigned int FillFlags(unsigned int flags)
```

Comment:
Would a better interface be to take a reference and update it in place?

Reply:
maybe not, since we're using it like
`CheckTxScripts(... FillFlags(flags) ...)`
So doing it in place would mean we need an extra line there